src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -55,25 +55,34 @@ public void storeMessage(Message message) {
    * @return a list of messages posted by the user, or empty list if user has never posted a
    *     message. List is sorted by time descending.
    */
-  public List<Message> getMessages(String recipient) {
+  public List<Message> getMessages(String name, String type) {
     List<Message> messages = new ArrayList<>();
+    Query query = (name != null) ? new Query("Message")
+                                    .setFilter(new Query.FilterPredicate(type, FilterOperator.EQUAL, name))
+                                    .addSort("timestamp", SortDirection.DESCENDING) :
+                                  new Query("Message")
+                                    .addSort("timestamp", SortDirection.DESCENDING);
 
-    Query query =
-        new Query("Message")
-            .setFilter(new Query.FilterPredicate("recipient",
-              FilterOperator.EQUAL, recipient))
-            .addSort("timestamp", SortDirection.DESCENDING);
     PreparedQuery results = datastore.prepare(query);
     for (Entity entity : results.asIterable()) {
       try {
         String idString = entity.getKey().getName();
         UUID id = UUID.fromString(idString);
-        String user = (String) entity.getProperty("user");
-
         String text = (String) entity.getProperty("text");
         long timestamp = (long) entity.getProperty("timestamp");
+        Message message = null;
+
+        if (type == "user") {
```

Comment:
String comparisons should be done with `constant.equals(variable)`, this checks value equality, not whether they are the same object.

Reply:
changed!